operator/internal/controller/pulumi/utils.go

```diff
@@ -41,3 +44,29 @@ func getGaugeValue(metric prometheus.Gauge) (float64, error) {
 	}
 	return m.Gauge.GetValue(), nil
 }
+
+// FinalizerAddedPredicate detects when a finalizer is added to an object.
+// It is used to suppress reconciliation when the stack controller adds its finalizer, which causes
+// a generation change that would otherwise trigger reconciliation.
+type FinalizerAddedPredicate struct {
+	predicate.Funcs
+}
```

Comment:
No need to add to this `utils.go` file, this can live in `stack_controller.go` since that's the only thing that uses it (or add a `predicate.go`).

```suggestion
type finalizerAddedPredicate struct {}
```